luci-app-babeld: add babeld overview

PolynomialDivision opened the pull request:
Add overview of babeld status, xroutes and routes.

PolynomialDivision commented:
![babeldapp](https://user-images.githubusercontent.com/10708466/102377748-4d204600-3fc5-11eb-83cb-520e17e32cc4.png)

PolynomialDivision commented:
Fixed Typo

feckert reviewed (changes requested):
Thanks for your contribution. See my comments-

* space vs. tabs in controller babeld.lua or change controller to menu.d json that would be better. Please have a look at https://github.com/openwrt/luci/blob/master/applications/luci-app-clamav/root/usr/share/luci/menu.d/luci-app-clamav.json
* Missing new line for babeld_network.lua

This is a status page. So why not move them to the status menu entry?
I could also imagine to add a polling to update this page?
But this can also be done afterwards.

I am not familiar with babld.
So what kind of command is this and where does the data come from?
https://github.com/openwrt/luci/pull/4664/files#diff-1dde564372ff1363073d8c79269ca9f57f5bf1b7527ec17e545d834cb2b5cf36R13

PolynomialDivision commented:
@feckert Thanks for your comments. Currently, I'm adding ubus bindings to babeld.
 That is why I maybe will refactor that again and will include your changes. ;)

PolynomialDivision commented:
>     * space vs. tabs in controller babeld.lua or change controller to menu.d json that would be better. Please have a look at https://github.com/openwrt/luci/blob/master/applications/luci-app-clamav/root/usr/share/luci/menu.d/luci-app-clamav.json

Changed.

>     * Missing new line for babeld_network.lua

Changed.

> This is a status page. So why not move them to the status menu entry?

What do u mean with that?

>  So what kind of command is this and where does the data come from?

From the local socket from babeld. Actually, I made a PR addding ubus bindings:
https://github.com/openwrt-routing/packages/pull/630

Since, this is still under discussion I would appreciate, if we could merge this app without ubus support, and then if the ubus bindings are there, I can rewrite. :)

PolynomialDivision commented:
@feckert Sry, just saw that code formating is not correct. I will change that.

PolynomialDivision commented:
Okay. Tested everything and fixed code format. I'm finished. :)

feckert commented on applications/luci-app-babeld/Makefile:
```diff
@@ -0,0 +1,10 @@
+include $(TOPDIR)/rules.mk
+
+LUCI_TITLE:=LuCI support for babeld
+LUCI_DEPENDS:=+babeld +luci-compat +luci-lib-json
```
Do we need luci-lib-json?

feckert commented on applications/luci-app-babeld/luasrc/model/cbi/babeld.lua:
```diff
@@ -0,0 +1,126 @@
+m = Map("babeld", "Babeld Overview", translate("Babeld Overview"))
```
The translation is missing from the second argument in the function. This is the heading of the section. The question is then only do you need the third argument in the function. This is usually used to describe the section more precisely. Just have a look at the other map functions in lua in the repo.

feckert commented on applications/luci-app-babeld/root/usr/share/luci/menu.d/luci-app-babeld.json:
```diff
@@ -0,0 +1,13 @@
+{
+	"admin/services/babeld": {
```
I have mentioned this before but I don't think you understood me ( I remember that you asked about). I think the overview is better kept in the status menu tree for bable than in the configuration. You only show status information here, it is not configuration page. See this for example https://github.com/openwrt/luci/blob/be49059e4d4ce156cfdf1b0240b05072b495d882/applications/luci-app-wireguard/root/usr/share/luci/menu.d/luci-app-wireguard.json#L2

feckert commented on applications/luci-app-babeld/luasrc/model/cbi/babeld.lua:
```diff
@@ -0,0 +1,126 @@
+		<%
+		local utl = require "luci.util"
+		local v6_port = 33123
+			local dump = utl.exec("(echo dump | nc ::1 %d)" % v6_port)
```
There the indentation does not fit.
I believe that the echo before dump is not necessary.
Is the command really called dump?

PolynomialDivision replied:
Is the command really called dump?
 
 Yes. ;)

feckert commented on applications/luci-app-babeld/luasrc/model/cbi/babeld.lua:
```diff
@@ -0,0 +1,126 @@
+		</div>
+		
+		<%
+		xroutes = {}
```
local missing

feckert commented on applications/luci-app-babeld/luasrc/model/cbi/babeld.lua:
```diff
@@ -0,0 +1,126 @@
+		</div>
+		
+		<%
+		routes = {}
```
local missing

feckert commented on applications/luci-app-babeld/luasrc/model/cbi/babeld.lua:
```diff
@@ -0,0 +1,126 @@
+				<%
+				for key,route in ipairs(routes) do
+					local route_sep = string.split(route," ")
+				%>
```
If the variable is not defined and has no content, do we not want to display `-`?
I am not familiar with bable, so I don't know if this can happen when the information gets dumped

PolynomialDivision replied:
that should not happen

feckert commented:
> Okay. Tested everything and fixed code format. I'm finished. :)

Looks good so far. Except for my comments.

PolynomialDivision commented:
Compile again and tested! :D Thanks a lot. I have to update my other luci apps accordginly to this. ;)

PolynomialDivision commented:
![statusbabeld cleaned](https://user-images.githubusercontent.com/10708466/102689316-6ad7f000-41fd-11eb-9cc0-544c0db74797.png)

feckert commented on applications/luci-app-babeld/luasrc/model/cbi/babeld.lua:
```diff
@@ -0,0 +1,126 @@
+		<%
+		local utl = require "luci.util"
+		local v6_port = 33123
+		local dump = utl.exec("(echo dump | nc ::1 %d)" % v6_port)
```
You still owe me an answer. Is the `echo` necessary?

PolynomialDivision replied:
Yes. If I remove it, the output is not correct? Or is there another way? I copied that from luci-app-olsr ;)

PolynomialDivision replied:
So I tested it. And it did not work.

feckert replied:
I think you need to redirect stdout or stderr.
I think the echo is not necessary here.
But as I said, I can't try it out.
You can write the difference in a file.

But that should not bother use. It's just an extra fork that I think is not neccessary.
Maybe we should write a FIXME comment above it.
If someone has the time and the inclination, they can take another look at it.

PolynomialDivision replied:
Added fixme comment. I will rewrite that when we have IPC via ubus merged. :)

PolynomialDivision commented:
I can not klick on request re-view. ;)
So I mention you again @feckert .

Thanks for the quick responses and the nice code review. :) This will not be the last commit to the babeld app. :)

feckert commented:
Thanks
